SCP-2663 Teach the Plutus Tx compiler to add profiling trace statements

thealmarty opened the pull request:
<!--
Here are some checklists you may like to use. Use your judgement.

This is just a checklist, all the normative suggestions are covered in more detail in CONTRIBUTING.
-->
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,21 @@ hoistExpr var t =
+                        let tInText = T.pack (show t'')
+                            -- Put "exiting" in the body
+                            traceInside (LamAbs () n x body) = LamAbs () n x (traceInside body)
+                            traceInside e                    = mkTrace ty ("exiting"<> tInText) e
```
this is going to be the wrong type. If we have
```
f :: a -> b -> c
f = \a -> \b -> body
```
then `ty` is `a -> b -> c`, but you want to have `trace @c "exiting f" body` on the inside, so you need to strip off bits of the type too.

michaelpj replied:
```
traceInside :: Term -> Type -> Term
traceInside (LamAbs n t body) (TyFun dom cod) = LamAbs n t (traceInside body cod)
traceInside (LamAbs{}) _ = error "no"
traceInside e ty = mkTrace ty ...
```
What happens with functions like:
```
foo = \a -> \b -> ...
foo a b = 
```

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,21 @@ hoistExpr var t =
+                            traceInside (LamAbs () n x body) = LamAbs () n x (traceInside body)
+                            traceInside e                    = mkTrace ty ("exiting"<> tInText) e
+                        return $
+                            mkTrace
```
This isn't quite right, I think. This puts the trace around the whole lambda, so it will trigger when we evaluate the lambda itself, not when we apply it. I think we need to do what I wrote in the ticket, something like
```
\a -> \b -> (trace @(() -> c) "entering f" (\() -> trace @c "exiting f" body) ())
```

michaelpj replied:
```
\a -> \b -> FORCE (trace "entering f" (DELAY (trace "exiting f" body)))
```

michaelpj replied:
```
let f = x +1
in f 1 + f 2
```

thealmarty commented:
- Added a profiling option in `-fplugin-opt`.
- Added an executable `profile` for profiling. When run, each program being profiled is written to `plutus-tx-plugin/executables/profile/`.
- The output files include timestamps and the name at each step. The files are to be processed to a flamegraph-like format in a follow up PR (see SCP-2365).

The executable runs fine locally. But I'm getting a `Plugins require -fno-external-interpreter` error on the CI. Otherwise this is ready for review.

michaelpj commented on plutus-tx-plugin/executables/profile/Main.hs:
```diff
@@ -0,0 +1,101 @@
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:debug-context #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:profile-all #-}
+
+-- | Executable for profiling. Add the program you want to profile here.
```
I think we should be able to do this as a test, no? Can't we just have a test module with `profile-all` where we have golden tests with the profile output?

michaelpj replied:
Hmm, but it'll include timestamps. So we need a "test time" emitter that emits some kind of deterministic time. Maybe we could have one with an internal counter and it just increments it every time it emits something?

thealmarty replied:
Yes, with the timestamps the golden tests will fail. If we want to make sure the programs compile correctly we should do the golden tests separately? Like the existing infrastructure in `test`. 

Yes we can have one with a counter. However, in that case my script that will turn these files to flamegraphs won't have as much detail on how much time has passed. 

Or are you suggesting we have both? I was assuming we know these compile fine and we just need to do profiling.

michaelpj replied:
Well, I think we want some tests that the profiling machinery works properly, and for those the counter one would work. i.e. I'm suggesting we use the counter version _only_ in testing. Because there we mostly care that the sequencing is working. 

For actually playing with examples it maybe does make sense to have an executable, yeah.

thealmarty replied:
I see. Makes sense. I can do that.

thealmarty replied:
OK I think I've done it? Please see [this commit](https://github.com/input-output-hk/plutus/commit/917da1bc064f31b9f7c8ea01a7f77fd1aeff79d0). Are the golden files looking right to you?

michaelpj replied:
Nice! Having done all that, I've realised that.... well, the counter is irrelevant, right? It just goes up. We already get the ordering we need from the trace statements anyway! So we could have just used the non-timestamping trace emitter :facepalm: sorry for the wild goose chase!

thealmarty replied:
No worries! You've made it very easy to add another emitter! :) I'll switch to the log one and remove this emitter.

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -422,6 +431,43 @@ hoistExpr var t =
+    -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+traceInside varName lamName (LamAbs () n t body) (PLC.TyFun () _dom cod) =
+    -- when t = \x -> body, => \x -> traceInside body
+    LamAbs () n t (traceInside varName lamName body cod)
```
nit: since `varName` and `lamName` are constant until later, I'd probably write this with a helper function. Something like:
```
traceInside varName lamName = go
  where 
    go (LamAbs ...) (PLC.TyFun ...) = ...
```

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -422,6 +431,43 @@ hoistExpr var t =
+                (PLC.TyFun () defaultUnitTy ty) -- ()-> ty
+                ("entering " <> varName)
+                -- \() -> trace @c "exiting f" e
+                (LamAbs () lamName defaultUnitTy (mkTrace ty ("exiting "<>varName) e)))
```
I'd probably just make this function monadic so you can generate the name here, but it's also fine to pass it in.

thealmarty replied:
As a FPer, I tend to avoid monads when I can. ;)

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -422,6 +431,43 @@ hoistExpr var t =
+        defaultUnit = PIR.Constant () (PLC.someValueOf PLC.DefaultUniUnit ())
+    in
+    --(trace @(() -> c) "entering f" (\() -> trace @c "exiting f" body) ())
+        PIR.Apply
```
looks right to me!

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,16 @@ hoistExpr var t =
+                        let ty = PLC._varDeclType var'
+                            varName = T.pack $ show $ PLC._varDeclName var'
+                        t'' <- compileExpr t
+                        lamName <- PLC.freshName "lamName"
```
I often call these unit arguments "thunk" or "dead" or something. I kind of want to make `_` a valid name so we can use that!

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,16 @@ hoistExpr var t =
+                            varName = T.pack $ show $ PLC._varDeclName var'
+                        t'' <- compileExpr t
+                        lamName <- PLC.freshName "lamName"
+                        return $
```
style nit: we tend to use `pure` rather than `return`.

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,16 @@ hoistExpr var t =
 
-                t' <- compileExpr t
+                CompileContext {ccOpts=profileOpts} <- ask
+                t' <- do
```
redundant do

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -405,7 +405,16 @@ hoistExpr var t =
+                t' <- do
+                    if coProfile profileOpts==All then do
+                        let ty = PLC._varDeclType var'
+                            varName = T.pack $ show $ PLC._varDeclName var'
```
I think you could pass in the `Name` and let `traceInside` decide what to do with it. Not a big deal in this instance, but it makes more sense to me that the user of `traceInside` knows it needs a `Name`, and it's only `traceInside` itself that knows that it's going to e.g. print it so it can stick in a string.

michaelpj reviewed (approved):
Looks plausible!

ak3n commented on plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs:
```diff
@@ -43,6 +45,12 @@ data CompileContext uni fun = CompileContext {
     }
 
+-- | Profiling options. @All@ profiles everything. @None@ is the default.
+data ProfileOpts =
+    All -- set this with -fplugin-opt PlutusTx.Plugin:profile-all
+    | None
+    deriving (Eq)
```
I guess it supposed to have more options later? Otherwise newtype over `Bool` might fit better.

thealmarty replied:
Yes, there may be more options later.

ak3n commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -422,6 +431,46 @@ hoistExpr var t =
+        (PIR.TyInst () (PIR.Builtin () PLC.Trace) ty)
+        [PLC.mkConstant () str, v]
+
+traceInside ::
```
I think it's worth to add a comment here.

michaelpj commented on plutus-tx-plugin/test/Plugin/Profiling/addInt.plc.golden:
```diff
@@ -0,0 +1,2 @@
+[ [0] entering Name {nameString = "addInt", nameUnique = Unique {unUnique = 81}}
```
Okay, we definitely don't want to print the names like this! Use one of the pretty-printing functions

michaelpj replied:
But it looks like the entering and exiting lines are in the right order, hooray!

thealmarty replied:
OK I think the golden files and the output files of the executable look better now.

michaelpj reviewed (commented):
Tests look great! It might be nice to have some more which have nested calls? At the moment there's only `fib`.

thealmarty commented:
> Tests look great! It might be nice to have some more which have nested calls? At the moment there's only `fib`.

Added `fact` [here](https://github.com/input-output-hk/plutus/pull/3779/commits/335d2e504c9a7c18fb256f55b5e05aba7d3fa68f).

thealmarty commented:
Other than the CI error, I think this is ready to go. Any idea about the CI error?

michaelpj commented:
We have some builds that do windows cross-compilation, but this won't work with the plugin. Your new component does need the plugin, so we need to tell Nix not to try building it when doing cross-compilation.

You can do it by adding it to the list here: https://github.com/input-output-hk/plutus/blob/master/nix/pkgs/haskell/haskell.nix#L80
In fact, I think it would indeed be fine to mark the entire `plutus-tx-plugin` package as unbuildable, so you can do it in exactly the same way as the others there.
